Add spherical sky region classes

sedonaprice opened the pull request:
**TL;DR: Proposed spherical regions classes (implemented, with demos); if greenlit, need to implement test suite, docs before this would be ready to mark as a "non-draft" pull request.  I would be very grateful for initial feedback by ~mid Sept (09/15).**

Hello!

First, a huge thank you to regions' developers, for making this extremely useful package available!

I'd like to propose an implementation of spherical sky regions, complementing the planar pixel and (implicitly planar) sky region classes. (I've seen a number of issues / past discussions on this topic and how it might be implemented, but my understanding is this development has not yet been done.)

I've discussed these proposed classes with @eteq, and I'm keen to get feedback from other key folks (gleaned from past issue discussions --- @pllim, @larrybradley,  @keflavich, and of course anyone else!)

Some background: I originally developed these classes as part of my work at STScI. Specifically, I needed "helper class" functionality to help demonstrate how to specify a selection region in one coordinate frame, and then transform to a different frame (and from there, construct a database query in new frame).  The infrastructure and general nature of this problem led me to first develop these classes within the regions framework -- and I hope these classes could be released to the broader astronomical community as part of the regions package.  **If these classes could be included in regions, this would be an ideal way of providing this spherical selection & transformation functionality needed as part of the Roman Research Nexus example user notebooks.**

I've put together a demo jupyter notebook walking through the core functionality (including plot demos) here: https://github.com/sedonaprice/regions/blob/add-spherical-regions-with-demo-NB/demo_spherical_regions.ipynb

**I have marked this as a draft pull request** because I have "tested" this code in a few demos (including the one linked above), but some work is outstanding as follows ***(pending a green light)***:
- ~Test suite for `SphericalSkyRegion` classes~ (complete)
- ~Create documentation pages + demos for the new functionality (and edit other doc pages as necessary) --- based on the below demo, but chunked out into separate topical entries.~ (complete
- ~Decision on handling of serialization for spherical regions (keyword switch to use spherical?), or add note to docs that serialization will default read in planar `SkyRegion` classes only.~  No serialization support for now / for consideration in a future PR. 
- ~Decision on whether or not (and how) to implement planar -> spherical conversion if boundary distortions are to be included.~ For consideration in a future PR.


**If folks think these classes would be a good candidate for inclusion, I will implement the outstanding test suite and documentation** (and other code additions, pending discussion on the points noted above).


**Related issues:**
These new classes would address the following issues:

* fix https://github.com/astropy/regions/issues/276
* fix https://github.com/astropy/regions/issues/303

and also help to achieve the aims in:

* https://github.com/astropy/regions/issues/564

An implementation (to be done) supporting conversion from planar to spherical accounting for boundary distortions would also address https://github.com/astropy/regions/issues/217  (I have ideas on how to implement this, based on the spherical region class discretization).

* (In the future, a solid angle property could be implemented in the SphericalSkyRegion classes, addressing https://github.com/astropy/regions/issues/434)


Thanks, and looking forward to any feedback!  I'd be very grateful for initial feedback by mid Sept (09/15), so I can work on next steps (either on the outstanding tasks, or figuring out some other way of enabling the functionality for the RRN example notebooks).

### TODO

- [x] Remove demo notebook.
- [x] Add proper documentation.
- [x] Add tests.
- [x] Squash out demo notebook in the commits or make sure maintainer use "Squash and Merge" button when merging.

sedonaprice committed:
Initial commit of spherical sky regions

Includes the initial basic spherical sky regions
    - SphericalSkyRegion
    - CircleSphericalSkyRegion
    - CompoundSphericalSkyRegion

sedonaprice committed:
Add spherical polygon, boundary discretization

- Implement PolygonSphericalSkyRegion
- Add a method discretize_boundary() to SphericalSkyRegion classes (for sampling along the boundary, for spherical->planar transformations that account for boundary distortions, and also for plotting)
- Implement spherical region calculation helper functions

sedonaprice committed:
Add bounding lonlat

- Add bounding_lonlat() methods to SphericalSkyRegion classes
- Implement spherical region calculation helper functions

sedonaprice committed:
Add spherical circle annulus

sedonaprice committed:
Add spherical lune

sedonaprice committed:
Add spherical range region

sedonaprice committed:
Add `to_spherical_sky()` to PixelRegion, SkyRegion classes

sedonaprice committed:
Codestyle fixes

sedonaprice commented:
Definitely, will chunk it out!

That sounds very reasonable as far as serialization and distortion (for planar -> spherical).

I'll note for completeness that I did implement distortion for spherical->planar (which I'd argue is good to keep in this PR and not split out; but perhaps I'm reading too much into your comment).

keflavich commented:
No need to split that out, imo.  I meant this bullet: "Decision on whether or not (and how) to implement planar -> spherical conversion if boundary distortions are to be included." should be a separate issue

pllim commented on .pre-commit-config.yaml:
```diff
@@ -8,6 +8,7 @@ repos:
     hooks:
       - id: check-added-large-files
         # Prevent giant files from being committed.
+        exclude: "demo_spherical_regions.ipynb"
```
```suggestion
```

sedonaprice replied:
Done

pllim commented on .pre-commit-config.yaml:
```diff
@@ -59,6 +60,7 @@ repos:
     hooks:
       - id: ruff
         args: ["--fix", "--show-fixes"]
+        exclude: "demo_spherical_regions.ipynb"
```
```suggestion
```

sedonaprice replied:
Done

pllim commented on .pre-commit-config.yaml:
```diff
@@ -89,6 +91,7 @@ repos:
         args: ["--write-changes"]
         additional_dependencies:
           - tomli
+        exclude: "demo_spherical_regions.ipynb"
```
```suggestion
```

sedonaprice replied:
Done

pllim commented on demo_spherical_regions.ipynb:
Move this elsewhere or break it into proper docs, as discussed.

sedonaprice replied:
Removed demo notebook from this pull-request branch; will break out into proper docs later.

sedonaprice replied:
Squashed.

pllim reviewed (dismissed):
I don't see any tests. Do you plan to add them?

I can't really do any meaningful review of the contents but maybe @mcara (maintainer of `spherical-geometry`) and @perrygreenfield can, or point to people who can.

Thanks!

sedonaprice commented:
(Demo notebook + precommit changes squashed out of this branch)

sedonaprice committed:
Implement test suite for spherical regions

- Add spherical tests to test_api.py
- Add test_compound.py
- Add test_whole_sky.py
- Add circle shape tests to test_circle.py
- Add annulus shape tests to test_annulus.py
- Add lune shape tests in test_lune.py
- Add range shape tests in test_range.py
- Add polygon shape tests to test_polygon.py

- Fix bugs in codebase & streamlining:
    - undeclared method in WholeSphericalSkyRegion
    - missing return statement in to_spherical_sky() for CircleAnnulusSkyRegion
    - add alternative "avg" centroid handling for spherical polygon; up for discussion on definitions + method choices
    - fix validation for TwoValAngleorNone
    - Add meta, visual to WholeSphericalSkyRegion for compatibility
    - Fix combound bounding_lonlat property
    - Fix logic error in lune  bounding_lonlat (vertex/ices on pole)
    - Fix internal unit formatting issue for centroid derived from cross product (radians vs other coordinate convention)
    - Bugfixes in edge cases for annulus logic in bounding_lonlat_poles_processing()
    - Streamline range methods/properties, removing unused & orphaned statements
    - Modify ComplexSphericalSkyRegion.frame property to avoid redefining in range
    - Fix range bugs: case of nverts=3 (touches pole); pole wrapping logic for defining latitude boundary shape (simplify and center on S pole in some cases); transformation when only lon or lat set; bounding circle only lon or lat bound synatx
    - Fix bug in ComplexSphericalSkyRegion.copy()
    - Update SphericalSkyRegion.frame attribute error message
    - Add explicit PolygonSphericalSkyRegion.centroid_avg, centroid_mindist properties for clarity

astrofrog reviewed (commented):
I've had a chance to take a look at this, and while I am not one of the regions maintainers, I think it would be great to include this so I would encourage you to proceed with writing the docs etc. This is very useful functionality! 💯 

I agree with @keflavich that we should worry about serialization after this PR though, because I think the main code/functionality is needed irrespective of whether we find ways to serialize it to some of the formats where there might be ambiguities.

At this point, the only thing I am not sure about API wise is the transformation of sky regions to other systems. In particular, while it's cool that this works:

```
>>> lonlat_range_transf = lonlat_range.transform_to("icrs")
>>> lonlat_range_transf
<RangeSphericalSkyRegion(
frame=<class 'astropy.coordinates.builtin_frames.icrs.ICRS'>,
longitude_bounds=<LuneSphericalSkyRegion(center_gc1=<SkyCoord (ICRS): (ra, dec) in deg
    (288.42757587, 10.72370325)>, center_gc2=<SkyCoord (ICRS): (ra, dec) in deg
    (217.98010947, -60.49575721)>)>,
latitude_bounds=<CircleAnnulusSphericalSkyRegion(center=<SkyCoord (ICRS): (ra, dec) in deg
    (192.85947789, 27.12825241)>, inner_radius=45.0 deg, outer_radius=90.0 deg)>
)>
```

It looks pretty complex (in particular the use of nested regions) and I'm not sure I understand the benefit of this. If someone accesses the vertices, they can always convert these between different frames, but I'm not sure that it makes sense to transform the whole spherical sky region. I guess fundamentally, what is the value in allowing sky regions to be transformed between frames since they are still fundamentally the same region? For instance, ``contains`` should work regardless of frames, and if someone does ``to_pixel`` for plotting they have to pass a WCS anyway. What use cases require ``transform_to``?

sedonaprice commented:
Hi @astrofrog --

For the specific case above ("range"), the issue with just using the vertices is that a lon/lat "range" is bounded by great circles along the lines of constant longitude, but **small or great** circles along the lines of constant latitude.  A spherical polygon defined with the same vertices would end up "bulging out" past the lines of constant latitude, so isn't equivalent.

I had added the "transforms_to()", including the nested regions, because circles transform neatly (as long as it's a spherical to spherical tranformation).  Thus, if a region can be represented as the appropriate circles (for polygons, indeed the transformation boils down to "transform the vertices and then remake the polygon, because great circles' centers transform easily; for small circles, again only the center transforms, the radius stays the same).

As I've implemented it, "contains()" actually does work regardless of frame (because the SkyCoord to be evaluated will be transformed to the region frame, if necessary)

The "transforms_to()" addresses the issue of "a user can define a region in their preferred non-ICRS frame, but needs to query against a database that only knows about ICRS" -- by transforming the region to ICRS, it's then possible to do bounding lon/lat in ICRS for search constraints, or a bounding circle (though of course the bounding circle transforms "easily".

(Definitely plotting requires "to_pixel" and a WCS, but the use case this was implemented to address is database/table searches with mismatches in coordinate frame.)

sedonaprice committed:
Correct docstring syntax for sphinx docs

sedonaprice committed:
Initial docs edits

- Getting started
- New index outline
- Placeholder files for new sections

sedonaprice committed:
Landing page example

sedonaprice committed:
Changes to shapes doc page

- Also explicitly change spherical regions write, serialize methods to raise NotImplementedError (as this handling is not yet implemented in the registry).

sedonaprice committed:
Edited contains docs to include spherical regions

sedonaprice committed:
Updated compound docs to add spherical regions

sedonaprice committed:
Added spherical example to plotting doc page

sedonaprice committed:
Spherical region frame transformation doc

- Add Spherical range to shapes docs (missed previously)
- Change representation to only print frame name, if it's defined
- Misc other doc plot improvements and bugfixes

sedonaprice committed:
Spherical region bounding info docs

- Misc docs plotting improvements

sedonaprice committed:
Bugfixes, update tests

- Fix documentation syntax bugs
- Update tests to reflect new ComplexSphericalSkyRegion representation of coordinate frame just by its name (if defined)
- Fix missing logical case in ComplexSphericaSkyRegion repr of the frame by name.

sedonaprice committed:
Add FLOAT_CMP to docs for testing

pllim commented:
I don't maintain this package, so I am not qualified to review. Sorry and good luck!

astrofrog commented:
@sedonaprice - I should have time to review this next week, could you rebase though to resolve the conflict?

sedonaprice committed:
Apply changes from #628 to CompoundSphericalSkyRegion

astrofrog commented on regions/shapes/tests/test_circle.py:
```diff
@@ -59,6 +63,24 @@ def test_pix_sky_roundtrip(self):
         assert reg_new.meta['text'] != self.reg.meta['text']
         assert reg_new.visual['color'] != self.reg.visual['color']
 
+    def test_sph_transformation(self, wcs):
```
This should hopefully be a simple search-and-replace job, but for clarity I wonder if you should replace ``test_sph_transformation`` by ``test_to_spherical_sky``

astrofrog commented on regions/shapes/tests/test_polygon.py:
```diff
@@ -5,18 +5,27 @@
+from regions.tests.helpers import assert_skycoord_allclose, make_simple_wcs
+
+
+def get_wcs():
```
Is this used anywhere? I didn't see it at first glance. If it is used, would it make more sense to have it be a fixture?

astrofrog commented on regions/shapes/range.py:
```diff
@@ -0,0 +1,812 @@
+    # --- only a concern for RangeSphericalSkyRegion....
+
+    @property
+    def __longitude_bounds(self):
```
I would just use a single underscore here (and there may be other ones, but I'l only comment here)

sedonaprice replied:
Here (and elsewhere for the latitude bounds), I was trying to do the following:
`_longitude_bounds` is set as an attribute. `longitude_bounds` is a property that returns the pre-set attribute `_longitude_bounds` if it is set, but if `_longitude_bounds` is None, then derive the longitude bounds on the fly through `__longitude_bounds`.

This chain was implemented because for Range instances that have changed coordinate frame, the longitude/latitude bounds must be passed when creating the instance, as there are no "ranges" stored (and thus the bounds cannot be derived on the fly) as it's no longer in the frame where the original range definitions make sense.

In retrospect, this is more obfuscated than it needs to be. I could change `__longitude_bounds` to a private method `_derive_longitude_bounds()`. This would make the purpose of the internal method more clear.

astrofrog commented on regions/shapes/whole_sky.py:
```diff
@@ -0,0 +1,74 @@
+        return None
+
+    def transform_to(self, frame):
+        # TODO: handle offset origin transformations
```
Do we need to if it's the whole sky?

sedonaprice replied:
Reading this again, this might relate to the "TODO" comment --- which is indeed not necessary here. This comment has been removed now.

If removing the comment resolves this point, this can be marked as resolved.

astrofrog reviewed (commented):
Overall this looks good, but one remaining big picture comment/question in addition to a few in-line comments below: are we expecting that some of the to_spherical_sky operations that are currently NotImplementedError might be implemented in future? My understanding is that some of them will never be implemented because they don't make sense/can't be done. The Python [docs](https://docs.python.org/3/library/exceptions.html#NotImplementedError) say about ``NotImplementedError``:

_Note It should not be used to indicate that an operator or method is not meant to be supported at all – in that case either leave the operator / method undefined or, if a subclass, set it to None._

Basically we should not be using ``NotImplementedError`` and instead should just not add the method where it doesn't make sense. In addition, if it can't be implemented for ``include_boundary_distortions=True``, I would suggest either removing that keyword argument altogether, or raising a ``ValueError`` if it's set to True.

Aside from this, please add an entry to the changelog.

sedonaprice committed:
Correct in-line comments

Correct in-line comments regarding transformation between shapes with discretization for cases requiring future work: these involve spherical to planar (or vice versa) transformation, not spherical to cylindrical.

sedonaprice committed:
Rename test to spherical sky

Rename tests `test_sph_transformation` to `test_spherical_sky`

sedonaprice committed:
Remove unused function in test_polygon.py

sedonaprice committed:
Fix typo

sedonaprice commented on docs/getting_started.rst:
```diff
@@ -236,3 +244,67 @@ You can access its properties via attributes:
+Spherical Sky Regions
+---------------------
+
+Spherical sky regions are defined using celescial coordinates,
```
Thanks -- fixed!

sedonaprice committed:
Split compound examples for smaller code chunks

sedonaprice commented on docs/compound.rst:
```diff
@@ -185,3 +216,28 @@ Example Illustrating Compound Regions
+
+    ax.set_xlim(-0.5, dataset.config['shape'][1] - 0.5)
+    ax.set_ylim(-0.5, dataset.config['shape'][0] - 0.5)
+    ax.set_title("SphericalSkyRegions")
```
Split into two; temporary example updated: https://sedonaprice.github.io/regions-docs-tmp/compound.html

sedonaprice committed:
Change test expected failure handling

Change tests to use `with pytest.raises(NotImplementedError)` instead of `except NotImplementedError`

sedonaprice committed:
Change lon/lat derivation handling in range

Change on-the-fly lon/lat derivation:
Change properties `Range.__longitude_bounds` and `Range.__latitude_bounds` to instead be internal methods `Range._derive_longitude_bounds()` and `Range._derive_latitude_bounds()`

sedonaprice committed:
Validate range lon/lat range & bound inputs

- Add to docstring that longitude values are expected to be within [0, 360] degrees, and that latitude values are expected to be within [-90, 90] degrees.
- Corrected docs examples to fit with this convention (this *could* be relaxed to allow for wrapping around longitude = 0, if such a change is desired in the future).
- Also change logic on longitude_bounds, latitude_bounds properties: Check if ranges for that coord are set before inferring "derive on the fly", as bounds of None is valid (eg, range of just Latitude or just Longitude).

sedonaprice commented on regions/shapes/range.py:
```diff
@@ -0,0 +1,812 @@
+        self._is_original_frame = kwargs.pop('_is_original_frame', True)
+
+        # TODO:
+        # Validate lon/lat range inputs, to ensure lat range has expected definition.
```
Implemented in c893145, for your review @astrofrog.

sedonaprice committed:
Update spherical <-> planar transform errors

- Change range spherical -> planar no distortions to raise ValueError (poorly defined)
- Change lune to raise a ValueError for no distortions (not defined, no analogous shape), and a NotImplementedErrors for distortions=True (as this can be defined, via boundary discretization)
- Update tests to catch ValueErrors instead of NotImplementedErrors where these are changed.

sedonaprice committed:
Added changelog entry

sedonaprice commented:
@astrofrog I've implemented:
- validation of lon/lat range inputs (and also the bounds inputs) in commit c893145
- An update to spherical <-> planar error messages, to reflect the logical changes noted in the table above. Namely, lune and range now raise `ValueError` when `include_boundary_distortion=False` (as these aren't analogous to planar shapes), and lune raises `NotImplementedError` for `include_boundary_distortion=True` (as discretizing the boundary into a polygon *does* have a planar equivalent, even though it would be rather irregular). (Commit f6a1aa8)

I've also added an entry to the changelog. If too verbose, I can reduce it to point users to docs pages with relevant details.

sedonaprice committed:
Fix assert vs error messages in range validation

sedonaprice committed:
Remove unnecessary comment from whole_sky.py

larrybradley committed:
Resolve conflict with main in test_api.py

Cherry pick of e2b43d7 into fork branch, with edits
to bring fork branch changes into consistency with the
new pytest parameterization usage.

Original commit:
e2b43d7: Fix pytest10 non-Collection iterable deprecation

sedonaprice committed:
Fix typo in plotting.rst

sedonaprice committed:
Fix deprecation in coord concatenation

Change spherical_helpers.py to use np.concatenate, following the astropy 7.2 "pending-deprecation" warning for the function astropy.coordinates.concatenate.

sedonaprice committed:
Bugfix polygon vertices ordering

Fixes an issue where the edge vertices ordering can be incorrectly flipped.

Change also removes unnecessary arguments from `discretize_all_edge_boundaries`, and propagates this change to methods that call this function.

sedonaprice commented:
Also @astrofrog --- I have also realized that the implementation of `PolygonSphericalSkyRegion.contains()` in this PR is only valid for **convex** polygons (computation simply uses the intersection of the great circles along which the edges lie on).

I have tested a generalized implementation (using astropy and numpy functions -- see branch [here](https://github.com/sedonaprice/regions/tree/add-spherical-regions-nonconvex-polygon)). In investigating solutions, I examined the code in [spherical_geometry](https://github.com/spacetelescope/spherical_geometry/tree/main) (particularly [here](https://github.com/spacetelescope/spherical_geometry/blob/afe09f20745dd2bc857a2249cef049e9b77e2761/spherical_geometry/great_circle_arc.py#L89)), but when coding it up I worked directly from the math in the source references ([here](https://web.archive.org/web/20160505193152/https://www.mathworks.com/matlabcentral/newsreader/view_thread/276271) and [here](https://en.wikipedia.org/wiki/Even%E2%80%93odd_rule); though I have not checked to see the behavior when a point lies _on_ the boundary --- which would make it inconsistent with circle contains excluding the boundary. Choices regarding edge contains conventions is probably best deferred to a separate discussion).

**Question:** Should I merge in this fix to enable general polygon support to this PR, or should that be the subject of a different PR after this one is approved?

Happy to discuss --- including thoughts from other maintainers or other relevant folks!

This of course relates to previous issue discussions regarding "should `spherical_geometry` be a dependency", but I figured it was worth exploring (a) a fix for the incorrect implementation in this PR, and (b) a no-dependency solution for at least the interim, or for use in case `spherical_geometry` was considered for addition as an optional dependency.

astrofrog commented on regions/core/compound.py:
```diff
@@ -252,5 +275,234 @@ def to_pixel(self, wcs):
+            lons2, lats2 = self.region2.bounding_lonlat
+            if (lons1 is None) | (lons2 is None):
+                return (None,
+                        Latitude([np.min([lats1[0].to(u.deg).value,
```
(same comment for all the instances below)

astrofrog commented on regions/shapes/range.py:
```diff
@@ -0,0 +1,893 @@
+            center_gc1, center_gc2, self.meta, self.visual
+        )
+
+    def _derive_latitute_bounds(self):
```
latitute -> latitude (and fix for any callers too)

astrofrog commented on regions/_utils/spherical_helpers.py:
```diff
@@ -0,0 +1,569 @@
+    vrepr = vertices.represent_as('spherical')
+
+    # lons_list = vrepr.lon
+    # lats_list = vrepr.lat
```
```suggestion
```

astrofrog commented on regions/shapes/whole_sky.py:
```diff
@@ -0,0 +1,72 @@
+        # Not defined
+        return None
+
+    def transform_to(self, frame):
```
Is this missing the ``merge_attributes`` kwarg from the base class?

astrofrog commented on regions/shapes/lune.py:
```diff
@@ -0,0 +1,240 @@
+        if discretize_kwargs is None:
+            discretize_kwargs = {}
+
+        if include_boundary_distortions:
```
This if statement is not needed since it'll always be ``True`` at this point?

astrofrog commented on regions/shapes/lune.py:
```diff
@@ -0,0 +1,240 @@
+        if include_boundary_distortions:
+            if wcs is None:
+                raise ValueError(
+                    "'wcs' must be set if 'include_boundary_distortions'=True"
```
There is some weird formatting here, why is ``include_boundary_distortions`` in quotes?

astrofrog commented on regions/_utils/spherical_helpers.py:
```diff
@@ -0,0 +1,569 @@
+            # UnitSpherical->SphericalRepresentation...
+            all_edge_bound_verts = SkyCoord(
+                SkyCoord(
+                    all_edge_bound_verts,
```
This is a bit confusing - here it looks like we are effectively doing ``SkyCoord(SkyCoord(SkyCoord(...``? I feel like there must be a better way 😅

astrofrog reviewed (changes requested):
This is looking good - I found a few issues below but once these are addressed, I'm happy with this, and I don't anticipate having time to review in full again.

One of the official maintainers should make a decision on merging this though (and whether or not they want to review themselves), this is beyond my authority.

sedonaprice committed:
Simplify/speed up with patthern `.to_value(UNIT)`

Change patterns of `.to(UNIT).value` to `.to_value(UNIT)`

sedonaprice committed:
Swap to use `np.minimum` directly on quantities

sedonaprice committed:
Correct discretization to use `circ_center`

sedonaprice committed:
Correct spelling of "latitude"

sedonaprice committed:
Correct `latitude_range` description

sedonaprice committed:
Correct docstring: second `lons_arr` -> `lats_arr`

sedonaprice committed:
Remove commented lines

sedonaprice committed:
Add `merge_attributes` kwarg to inheriting `transform_to`

sedonaprice committed:
Add tildes for links to package classes

sedonaprice committed:
Simplify discretization boundary SkyCoord creation

Remove unnecessary nested calls to SkyCoord

sedonaprice committed:
Remove unnecessary if statement in lune.py

sedonaprice committed:
Remove other unnecessary if statements from lune,range

sedonaprice committed:
Change include_bound_dist + no WCS error msg format

sedonaprice committed:
Simplify range bound transformation method

sedonaprice committed:
Common _validate_planar_spherical_transform() method

Migrated checking of valid planar <-> spherical transformation inputs to a common private `Regions._validate_planar_spherical_transform()` validation method.

sedonaprice commented:
@astrofrog Thank you for your thorough review, and your suggestions! I've implemented all changes (including other places in the code where equivalent changes could be made, where relevant).

@keflavich and @larrybradley --- as maintainers of `regions`, I wanted to check in about next steps for considering this for merging.

I wanted to flag again the following (which I noted on 9 Jan):

> I have also realized that the implementation of PolygonSphericalSkyRegion.contains() **as currently implemented in the branch for this PR** is only valid for convex polygons (computation simply uses the intersection of the great circles along which the edges lie on).

Possible solutions include:
1. Consider this PR code as-is, modifying the doc strings (and/or SphericalSkyPolygon class name) to note it's currently valid for complex polygons. Make modifications to ensure generalized spherical polygons are supported in a future PR.
2. Fix this issue, implementing generalized spherical polygons before completing this PR. (I worked out such a fix in a [separate branch](https://github.com/sedonaprice/regions/tree/add-spherical-regions-nonconvex-polygon).) Longer term, it might be worth considering adding an optional dependency to handle generalized polygon contains logic (e.g., `spherical_geometry`).

Happy to discuss how best to tackle this issue, and thank you again everyone for your time and extremely useful feedback!

astrofrog commented:
@sedonaprice - this will need a rebase, but once you do I'm happy to approve this assuming everything looks ok, and then @keflavich or @larrybradley should make final decision

lpsinger commented:
Just to add my :+1: to this idea and PR because I too have an application for which I am pretending that SkyRegions are spherical geometry and I would rather see real support (see https://m4opt.readthedocs.io/en/latest/guide/fov.html#supported-region-types).

sedonaprice committed:
Document convex-only spherical polygon limitation

Additionally, removed unnecessary "code commentary" in `CircleSphericalSkyRegion` and `PolygonSphericalSkyRegion` docstrings.

sedonaprice committed:
Update test: improved pixel <-> sky conversion

Following the improvements in PR #650 for pixel <-> sky conversions, the CircleSphericalSkyRegion to_pixel() test failed. Update the radius value, matching the updates for the planar sky transformation (as implemented in https://github.com/astropy/regions/pull/650/changes/78a2e3dae2798b8e658c9668dc8957234e42c62d)

sedonaprice committed:
Update shapes.rst for repr, wcs sky<->pix changes

Following changes in PRs #655, #650, update the shapes.rst docs to reflect
- Changes to value representation (truncation)
- Changes to transformed circle radius values (WCS improvements)

sedonaprice commented:
Thanks folks!

I have:
- Rebased against astropy:main
- Added documentation that spherical polygon only supports convex polygons (0df756e). **Suggestions of edits for additional clarity welcome!**
- Updated some failing tests and documentation tests (given changes in #650, #655).

@astrofrog --- the rebase looks good, and the test failures that were corrected are understandable as far as other changes that were made.

@keflavich @larrybradley -- please let me know if I should make additional clarifications to document the limitations for PolygonSphericalSkyRegions for now, as far as taking route (1).

sedonaprice committed:
Run pre-commit for linting

sedonaprice commented:
As an aside point:

Given #649, would there be interest in changing the spherical regions'  `SphericalSkyRegion.discretize_boundary()` methods to `SphericalSkyRegion.to_polygon()`, for parity with the new planar pixel and sky `to_polygon()` methods?

This would be a bit recursive as far as naming for `PolygonSphericalSkyRegion`, but discretizing/"extra polygonizing" a polygon's edges is key to capturing boundary distortions in planar<->spherical.... retaining the name `discretize_boundary()` would avoid this confusion, but would do effectively the same thing as `to_polygon()` with a different name.

(Alternatively an additional method `to_polygon()` could be added for the spherical shapes where that is logical, providing an alias of `discretize_boundary()`.

Not a critical item, but I wanted to raise the possibility pre-merge.

sedonaprice commented:
@keflavich @larrybradley I wanted to ping again about your high-level review, comments on how I addressed changes to take route 1 (document limitations of convex polygons for now), and if we could move ahead with merging the spherical regions --- thanks!

keflavich commented on docs/getting_started.rst:
```diff
@@ -236,3 +244,67 @@ You can access its properties via attributes:
+These distortions (implemented through discrete boundary sampling)
+capture the impact of the spherical-to-planar (or vice versa) projection.
+However, it is possible to ignore these distortions (e.g.,
+transforming a spherical circle to a planar circle).
```
Could we add an example here showing the difference - i.e., showing how a point could go from inside the spherical circle to outside the planar circle, and vice-versa? The more clear we can make the limitations of each approach, the better.

keflavich reviewed (approved):
I have a minor suggestion on something to add to the docs, but in my view, this is good to go and high-value - I'd like to see it merged soon.

keflavich commented:
also I'd be happy to add the `to_polygon` method, but I don't feel strongly about it.  It might be better understood as `to_projected_polygon`, but then you lose the symmetry advantage.  But maybe the mnemonic value is worth adding it?

larrybradley commented on regions/shapes/lune.py:
```diff
@@ -0,0 +1,239 @@
+        if not include_boundary_distortions:
+            raise ValueError(
+                'Invalid parameter: `include_boundary_distortions=False`!\n'
+                'Transforming range to planar pixel region is only possible by '
```
```suggestion
                'Transforming lune to planar pixel region is only possible by '
```

larrybradley commented on docs/spherical_bounding_regions.rst:
```diff
@@ -0,0 +1,256 @@
+
+    import matplotlib.pyplot as plt
+    from matplotlib.lines import Line2D
+
```
```suggestion
import numpy as np
```

`np` is used in the plot, but never imported.

larrybradley commented on regions/core/core.py:
```diff
@@ -426,6 +437,39 @@ def to_sky(self, wcs):
+            Ignored if boundary distortions not included.
+
+        include_boundary_distortions : bool, optional
+            If True, accounts for boundary boundary distortions in spherical to planar
```
```suggestion
            If True, accounts for boundary distortions in spherical to planar
```

larrybradley commented on regions/core/core.py:
```diff
@@ -653,3 +697,432 @@ def to_pixel(self, wcs):
+            Ignored if boundary distortions not included.
+
+        include_boundary_distortions : bool, optional
+            If True, accounts for boundary boundary distortions in spherical to planar
```
```suggestion
            If True, accounts for boundary distortions in spherical to planar
```

larrybradley commented on regions/core/core.py:
```diff
@@ -653,3 +697,432 @@ def to_pixel(self, wcs):
+            Ignored if boundary distortions not included.
+
+        include_boundary_distortions : bool, optional
+            If True, accounts for boundary boundary distortions in spherical to planar
```
```suggestion
            If True, accounts for boundary distortions in spherical to planar
```

larrybradley commented on regions/core/core.py:
```diff
@@ -653,3 +697,432 @@ def to_pixel(self, wcs):
+            Ignored if boundary distortions not included.
+
+        include_boundary_distortions : bool, optional
+            If True, accounts for boundary boundary distortions in spherical to planar
```
```suggestion
            If True, accounts for boundary distortions in spherical to planar
```

larrybradley commented on regions/shapes/lune.py:
```diff
@@ -0,0 +1,239 @@
+                "'wcs' must be set if `include_boundary_distortions=True`",
+            )
+        # Requires spherical to planar projection (from WCS) and discretization
+        raise NotImplementedError
```
Why is there a bunch of validation code that ends in `raise NotImplementedError`?  I think that code should be removed.  Also, this `to_pixel` method is called only by `Lune....to_sky`.  But that means the `to_sky` method never succeeds.  Should that simply `raise NotImplementedError` too?

sedonaprice replied:
I'd intended to implement this, but it slipped by. Will implement it now!

larrybradley reviewed (changes requested):
Thanks again, @sedonaprice. It's getting close. I have a few small changes.

For consistency, I do recommend adding a `to_polygon` method that delegates to `discretize_boundary`.   `discretize_boundary` uses `n_points`, but `to_polygon` uses `npoints`.  Let's go with `n_points` everywhere for consistency.  The `to_polygon` methods are recent additions and have not yet been in a release, so we don't need to deprecate `npoints`.

larrybradley commented:
@sedonaprice Not needed for this PR, but I think it would be good for a followup PR to add a check for input non-convex polygons in `PolygonSphericalSkyRegion` that raises an error.  I think this can silently return wrong answers for non-convex inputs (users may not read the docs).

larrybradley commented:
I'm also wondering if `n_points` should be 100 everywhere.  In a couple places, you have `n_points=10`.

larrybradley commented on docs/contains.rst:
```diff
@@ -60,3 +63,34 @@ Note that `regions.SkyRegion.contains` requires a WCS to be passed::
+
+For `~regions.SphericalSkyRegion` objects, checking whether point(s) are
+contained inside that region requires no other input --- since these
+regions are defined with a the spherical geometry, and not a projected geometry
```
```suggestion
regions are defined with a spherical geometry, and not a projected geometry
```

larrybradley commented on regions/shapes/lune.py:
```diff
@@ -0,0 +1,239 @@
+
+class LuneSphericalSkyRegion(SphericalSkyRegion):
+    """
+    Class for a spherical "lune", the intersection between to great
```
```suggestion
    Class for a spherical "lune", the intersection between two great
```

larrybradley commented on regions/shapes/circle.py:
```diff
@@ -92,6 +94,29 @@ def to_sky(self, wcs):
                                visual=self.visual.copy())
 
+    def to_spherical_sky(self, wcs=None, include_boundary_distortions=False,
+                         discretize_kwargs=None):
```
Looking at this PR in more depth, I do have a more substantial request.  Please remove the `discretize_kwargs` kwarg everywhere and instead replace it with `n_points=None`.  

1. `n_points=1000` is way less verbose than `discretize_kwargs={'n_points': 1000}`
2. `discretize_boundary` only accepts a single `n_points` keyword
3. A typo like `discretize_kwargs={'n_point': 1000}` passes silently (the keyword is ignored and there is no validation anywhere)
4. This change will eliminate a lot of boilerplate code
5. Making the `n_points` keyword explicit allows it be discovered in `help()` (and IDE autocompletes, etc.)

The docstring should be updated to note that the `n_points` keyword will have no effect unless `include_boundary_distortions=True`.

You could do something like this:
```python
def to_sky(self, wcs=None, include_boundary_distortions=False, n_points=None):
    ...
    if include_boundary_distortions:
        disc_kwargs = {} if n_points is None else {'n_points': n_points}
        disc_bound = self.discretize_boundary(**disc_kwargs)
```

sedonaprice commented:
Hi folks, thanks for the detailed comments! 

At glance I can implement basically everything relatively quickly. I'll also try to add the docs example of "spherical vs planar" circle in/outside a circle as @keflavich suggested, as that's not too time intensive. 

I'll go through these in the coming days and let folks know when things are ready for re-review!

@larrybradley re checking for non-convexity of input polygons: I'd definitely like to chat with you and others about how to actually implement that check, as I don't know a convenient algorithm for that off the top of my head. (Maybe checking with folks working on `spherical-geometry`).  

(Thanks also for the suggestion to remove `discretize_kwargs` --- I'd thought about that at one point after the development matured, but didn't follow through!)

larrybradley commented:
@sedonaprice For the concave (non-convex) check, perhaps something like comparing the convex hull area to the polygon area would work.  For a concave polygon, the convex hull area will be be larger than the polygon area.

The `spherical_geometry` `SphericalPolygon` class (https://spherical-geometry.readthedocs.io/en/latest/api/spherical_geometry.polygon.SphericalPolygon.html#spherical_geometry.polygon.SphericalPolygon) has `area` and `convex_hull` methods.  It's probably a good idea to input an interior point (`inside=`).
CC: @mcara for `spherical_geometry` advice

sedonaprice commented:
Sure -- but my understanding was there were past decisions to not have `spherical_geometry` be a (required) dependency of `regions`.  It might be more expedient to just a pure-pythonic spherical polygon fix that works for convex and non-convex (eg, as I prototyped in a separate branch, starting from the same algorithm as in `spherical_geometry`, but of course lacks quad precision on cross products).  

(I'd also argue that it might be good to permit users to not pass interior points, as eg users would not be able to simply pass vertices from an `s_region`.)

mcara commented:
@larrybradley My understanding is that these "spherical sky regions" (or polygons) are fundamentally planar polygons. To check that polygons are convex:
1. detect self-intersecting polygons. By definition they are not convex. AND
2. perform cross-product sign test: https://math.stackexchange.com/a/1745427 (pretty simple to implement in Python using `numpy`).

For polygons on the sphere, the analog on the sphere of method from `#2` above would be the half‑space sign test. 

With regard to the `inside` point: It is used to indicate the inside face of the spherical polygon. We are investigating the use of this in the `spherical_geometry`. At this moment, my preliminary "feeling" about it is that it introduces more confusion than it helps. So, at this moment, I cannot recommend introducing this parameter

sedonaprice commented:
Hi @mcara -- regarding "My understanding is that these "spherical sky regions" (or polygons) are fundamentally planar polygons.".  These spherical sky regions are not planar: the spherical polygons are sky coordinates joined by a great circle arcs, the spherical circles have a circumference defined by constant spherical projected distance from a sky coordinate, etc.  Could you clarify what you meant by that statement?

sedonaprice commented:
And thank you @mcara! A check for (1) self-intersection and (2) leveraging the half-space sign test would be fairly straightforward to implement.

For completeness as far as implementing a convexity check, @larrybradley, I will note that if this test was attached to the creation of a `SphericalSkyPolygon`, then in some cases when discretizing to transform from a spherical shape, like a circle, to a planar shape while accounting for the WCS transformation boundary distortions, the transformed, discretized boundary can be non-convex, so the full functionality of those transformations wouldn't be supported until arbitrary polygons are supported.

mcara commented:
> Could you clarify what you meant by that statement?

I am not familiar with regions package and I guess I did not read the getting started rst completely and I was thinking of "planar Sky Regions". In this case use "half‑space sign test".

With regard to `include_boundary_distortions` - from a quick look it seems this would throw a `NotImplemented` error. Is that correct? If true, I would suggest not having this parameter at all. Also, documentation about `include_boundary_distortions` mentions only projection distortions but I could not find any mention of instrument distortions. Are those ignored?

sedonaprice commented:
Thanks for the clarification!

For `include_boundary_distortions`, no, it doesn't throw a `NotImplementedError` in general. For `LuneSphericalSkyRegion` and `RangeSphericalSkyRegion` (aka an lon/lat bounding box), trying to transform to a planar shape with `include_boundary_distortions=False` will throw `NotImplementedError`, as these spherical shapes have no direct analogs --- for a lune, it's full "orange slice" and a canonical planar shape can't approximate it, and for range this ensure large lon/lat ranges aren't approximated as a rectangle, which is pretty far off.  Note that for `LuneSphericalSkyRegion` and `RangeSphericalSkyRegion` a transformation to a planar shape with `include_boundary_distortions=True` will work, as the polygonization enables it to be described on a plane.

By "projection distortions", I intended to imply "all the distortions happening to a boundary when mapping from spherical geometry to a planar projection as described by the WCS considered, or vice versa". This would indeed include instrument distortion if included in the WCS.  Very welcome to suggestions for a more clear way to phrase this --- would saying "WCS projection distortions" make this more clear?

mcara commented:
I use "projection distortion" to describe distortions arising from projecting a "straight" line (an arc) from a sphere to a plane. The "distortion" will depend on the projection used in the WCS. It is separate from *instrument* distortions usually described using SIP, tabular distortions, or TPV.

sedonaprice commented:
My understanding is in general the WCS for reduced images/mosaics as parsed by astropy.wcs would include both projection distortion and also components described by SIP.

I'll look back at the docstrings & docs to see if it makes sense to add "WCS distortion" or "distortion described by the WCS transformation" to clarify things!

sedonaprice committed:
Fix docs, docstring typos

sedonaprice committed:
Add missing import

sedonaprice committed:
Change default to `n_points=100` everywhere

sedonaprice committed:
Add lune GC center invert test

Also add short test annotation for test_range.py

sedonaprice committed:
Fix lune boundary discretization

Lune boundary discretization was in CCW order; invert to fix for now. Note a TODO to enable proper CW order checking of all polygons at some future point (likely a shared method used in many places to validate/re-order polygon vertices).

sedonaprice committed:
Discretize to total n_points, not per-edge

Previous implementation discretized to n_points per edge. Change to discretize to a total of n_points vertices, for consistency with to_polygon.

Update tests accordingly.

sedonaprice committed:
Apply meta, visual to polygon in `discretize_boundary()`

Apply a region's meta, visual to its the resulting PolygonSphericalSkyRegion when calling `discretize_boundary()`, for consistency with `to_polygon()`

sedonaprice committed:
Whole sky discretization not implemented test

Verify whole sky returns a NotImplementedError when calling `discretize_boundary()`

sedonaprice committed:
Implement `to_polygon()` method

Implement a `to_polygon()` method that delegates to `discretize_boundary()`, for consistency with the newly-implemented planar regions `to_polygon()` method.

sedonaprice committed:
Docs: more explicitly note "WCS distortion"

Update the docs to note the "boundary distortion" choice for spherical-to-planar and vice versa transformation captures all effects encoded in the WCS, not just "spherical-to-planar" projection effects.

sedonaprice committed:
Remove `discretize_kwargs` for `n_points=None`

Removing `discretize_kwargs` as a sky/pixel/spherical sky transformation input in favor of the streamlined `n_points=None` kwarg, as suggested/requested in https://github.com/astropy/regions/pull/618#discussion_r3151300518

sedonaprice committed:
Change WholeSky errors for undefined methods

Change WholeSphericalSkyRegions errors for undefined methods to ValueError, as these will never be implemented.

sedonaprice committed:
Implement lune distorted sph->planar

Implement the LuneSphericalSkyRegion transformations `to_sky()` and `to_pixel()` when `include_boundary_distortions=True`.

(Also update to use common transformation validation, and  correct some notes for possible future implementations on other shapes' transformation methods.)

sedonaprice committed:
Add docs demo of spherical vs planar circle

Add a short visual demonstration of the difference between spherical and planar circles to "Getting started".

Modifies the "Spherical Sky Regions" section to have a new "Spherical to planar region transformations" subsection.  The text discussing spherical-to-planar transformations is moved to that subsection, and  added a short discussion of the plot demo.

sedonaprice commented:
@larrybradley review comments addressed!
* Refactored to remove `discretize_kwargs` in favor of `n_points=None`
* Added `to_polygon()` method.
* Implemented `to_sky()`, `to_pixel()` for lune when `include_boundary_distortions=True`
* As requested by @keflavich, added a demo of spherical vs planar regions to getting_started.rst (slightly rearranging the spherical regions section for better flow). Rendered version here: https://sedonaprice.github.io/regions-docs-tmp/getting_started.html
* Based on @mcara's comments, modified the documentation wording to be more clear that the boundary distortion would capture all distortions encoded in the WCS.
* Misc small changes requested
* Other small improvements *(changing discretize to have n_points TOTAL, not per edge, for consistency with the planar shapes' methods; added some notes in tests; change WholeSky error for undefined methods to ValueError, not NotImplementedError, as these will never be implemented)*
* Rebased against main to pick up changes from #659. (New changes from this review from 639a0b1 onwards.)

keflavich commented:
I like the new example and graphic. 👍

larrybradley commented on regions/core/core.py:
```diff
@@ -6,12 +6,15 @@
 import astropy.units as u
 import numpy as np
 from astropy.coordinates import SkyCoord
+from astropy.coordinates.sky_coordinate_parsers import _get_frame_class
```
Sorry I missed this in the earlier review, but importing a private function from external package is generally considered unsafe as it could change without warning.  Is there a way use a public method instead?

sedonaprice replied:
I think the only practical way is to directly implement the same logic from as a stand-alone function: `astropy.coordinates.sky_coordinate_parsers._get_frame_class()` in regions ([astropy code](https://github.com/astropy/astropy/blob/4031ec7401d27745c545e56648da19270401e12b/astropy/coordinates/sky_coordinate_parsers.py#L40)).  It's not actually too heavy and relies mostly on the `astropy.coordinates.baseframe.frame_transform_graph` instance, but then this implement could be out of step with the up-to-date astropy implementation.

The problem is that the only public way of accessing this method is through a SkyCoord, but it's not possible to create a data-less SkyCoord, and a SkyCoord is not always available from spherical region inputs (specifically, for Range). And not all frames have the same "data shape" to use for a placeholder SkyCoord to access that function (e.g., `SkyCoord(0,0,unit='deg',frame='galactocentric')` fails).

@larrybradley I'll implement a port of that function, but I'd be happy to change to something else if there's some other clever solution.

(Of course if astropy.coordinates exposed this as a public function this would be another solution, but I can understand why that would be non-ideal.)

larrybradley replied:
Thanks!  Your stand-alone port looks fine here.  
Going forward, You could create an astropy issue to see if there is any interest to make that function public.

larrybradley commented on regions/core/core.py:
```diff
@@ -653,3 +698,435 @@ def to_pixel(self, wcs):
+            The world coordinate system transformation to use to convert
+            between sky and pixel coordinates. Required if transforming
+            with boundary distortions (if ``include_boundary_distortions`` is True).
+            Ignored if boundary distortions not included.
```
It appears that the `to_pixel` method *always* requires a `wcs` object.  For example, even with `include_boundary_conditions=False`, `CircleSphericalSkyRegion.to_pixel` always calls `self.to_sky().to_pixel(wcs)`.

sedonaprice replied:
Ah, good catch, will update docstring accordingly!

larrybradley replied:
Since `wcs` is always required for this method, I don't think it should be an optional keyword with a default of `None`.  I recommend that it be changed to a positional argument.

larrybradley commented on regions/core/core.py:
```diff
@@ -426,6 +437,40 @@ def to_sky(self, wcs):
+
+        n_points : int, optional
+            The number of polygon vertices for boundary discretization.
+            This keyword will have effect unless ``include_boundary_distortions=True``.
```
```suggestion
            This keyword will have no effect unless ``include_boundary_distortions=True``.
```

larrybradley commented on regions/core/core.py:
```diff
@@ -653,3 +698,435 @@ def to_pixel(self, wcs):
+
+        n_points : int, optional
+            The number of polygon vertices for boundary discretization.
+            This keyword will have effect unless ``include_boundary_distortions=True``.
```
```suggestion
            This keyword will have no effect unless ``include_boundary_distortions=True``.
```

larrybradley commented on regions/shapes/line.py:
```diff
@@ -84,6 +84,10 @@ def to_sky(self, wcs):
                              visual=self.visual.copy())
 
+    def to_spherical_sky(self, wcs=None, include_boundary_distortions=False,
+                         discretize_kwargs=None):
```
```suggestion
                         n_points=None):
```

sedonaprice replied:
And found other case; both fixed!

larrybradley commented on regions/shapes/line.py:
```diff
@@ -197,3 +201,7 @@ def to_pixel(self, wcs):
                                visual=self.visual.copy())
+
+    def to_spherical_sky(self, wcs=None, include_boundary_distortions=False,
+                         discretize_kwargs=None):
```
```suggestion
                         n_points=None):
```

larrybradley commented on docs/getting_started.rst:
```diff
@@ -236,3 +244,165 @@ You can access its properties via attributes:
+
+The example below demonstrates the difference between a spherical and
+a planar circle with the same center and radius. Boundary distortions
+are included when projection the spherical circle onto the plot.
```
```suggestion
are included when projecting the spherical circle onto the plot.
```

larrybradley reviewed (changes requested):
Thanks again, @sedonaprice.  I have a couple new comments about a private import and `to_pixel` needing `wcs` (along with a few typos).

sedonaprice committed:
Validate same frame for spherical compound regions

Add a validation check to ensure both regions of a compound region have the same frame.

sedonaprice committed:
Improve _validate_frame()

Separate _validate_frame() into separate logical pieces:
* Standardize the frame into an astropy frame class with _standardize_frame() (static method). Called for both standardization in _validate_frame(), and upon setup for `RangeSphericalSkyRegion`
* Only call _validate_frame() for frame transformations. Added an explicit check for whether the current frame can be transformed to the target frame.

sedonaprice committed:
Implement stand-alone port _get_frame_class

sedonaprice committed:
Update to_pixel() docstring

sedonaprice committed:
Update n_points descriptions

sedonaprice committed:
Remove `discretize_kwargs` from line.py

sedonaprice committed:
Fix typo in getting_started.rst

sedonaprice committed:
Rename to _validate_frame_transformation

Rename _validate_frame to _validate_frame_transformation to make method purpose more clear.

sedonaprice commented:
@larrybradley Thanks!
Comments addressed.

For the private method, I implemented a stand-alone port of `astropy.coordinates.sky_coordinate_parsers._get_frame_class()` (7a71ca0; see comment above), but can roll that back if you would prefer some other solution.

I also took the opportunity to streamline internal frame checks (regions.core.core.py, L791-816 now), splitting into separate standardization and validation methods. _(Previously the single method both standardized the frame input and held a placeholder to raise origin offset errors --- but this method was also being used to standardize the frame setup for `RangeSphericalSkyRegion`. Now the validation method is only called from `transform_to()`, and explicitly performs a check for whether the frame is transformable.)_

sedonaprice committed:
Make wcs a posarg in spherical region to_pixel()

Change spherical regions' to_pixel() method to take wcs as a positional argument, as wcs is always required for that method.

sedonaprice commented:
@larrybradley wcs now a positional argument in `SphericalSkyRegion.to_pixel()`!

larrybradley reviewed (approved):
@sedonaprice Many thanks for seeing this through! :tada:
